dot/parachain/dispute/queues.go

```diff
@@ -0,0 +1,242 @@
+package dispute
+
+import (
+	"bytes"
+	"sync"
+
+	parachain "github.com/ChainSafe/gossamer/dot/parachain/types"
+	"github.com/ChainSafe/gossamer/lib/common"
+
+	"github.com/pkg/errors"
+	"github.com/tidwall/btree"
+)
+
+// This file contains the types and methods for the queues used by the dispute coordinator
+// The implementation is based on parity's participation queue
+// Read more here: https://paritytech.github.io/polkadot/book/node/disputes/dispute-coordinator.html
+// https://github.com/paritytech/polkadot/blob/master/node/core/dispute-coordinator/src/participation/queues/mod.rs
+// It uses btree. Find it here: https://github.com/google/btree
+
+// TODO: Parity's implementation captures metrics for the queue. We should do the same.
+// However, I will not be implementing it right away. It will be picked up as a single task for the
+// entire dispute module https://github.com/ChainSafe/gossamer/issues/3313.
+
+// CandidateComparator comparator for ordering of disputes for candidate.
+type CandidateComparator struct {
+	relayParentBlockNumber *uint32
+	candidateHash          common.Hash
+}
+
+// ParticipationRequest a dispute participation request
+type ParticipationRequest struct {
+	candidateHash    common.Hash
+	candidateReceipt parachain.CandidateReceipt
+	session          parachain.SessionIndex
+	//TODO: requestTimer for metrics
+}
+
+// ParticipationItem implements btree.Item
+type ParticipationItem struct {
+	comparator CandidateComparator
+	request    *ParticipationRequest
+}
+
+func participationItemComparator(a, b any) bool {
+	pi1, pi2 := a.(*ParticipationItem), b.(*ParticipationItem)
+
+	if pi1.comparator.relayParentBlockNumber == nil && pi2.comparator.relayParentBlockNumber == nil {
```

Comment:
do we need a nil check on the comparator itself?

Reply:
comparator can never be nil. See https://github.com/ChainSafe/gossamer/blob/feat/k/participation-queue/dot/parachain/dispute/queues.go#L98